test/tests/functional/pbs_qrun.py

```diff
@@ -115,11 +115,13 @@ def test_qrun_hangs(self):
             else:
                 self.server.submit(j)
         self.logger.info("Submitted 500 jobs with different walltime")
-        now = int(time.time())
         self.server.manager(MGR_CMD_SET, SERVER,
                             {'scheduling': 'True'})
         self.server.manager(MGR_CMD_SET, SERVER,
                             {'scheduling': 'False'}, expect=True)
+        now = int(time.time())
+        self.scheduler.log_match("Leaving Scheduling Cycle", starttime=now)
+        now = int(time.time())
```

Comment:
if you are setting 'now' after ensuring that the scheduler is not in a sched cycle, then do you really need to do the log_match() above?

Reply:
Yes the log match for "Leaving Scheduling Cycle" is necessary. 
1. The first "now" am setting because we need the current cycle we just started that needs to complete, in sched log. Otherwise, even though in server the "scheduling" attribute is set false, but in sched log the end message of current cycle still not received.

2. For fast system, the first now timestamp can also be same as when we are starting the cycle. So when we start checking log_match() in parent process after fork, it matches with the 1st cycle starting log.
To overcome this, noticed the timestamp when we got the "Leaving Scheduling Cycle" message in sched log. Started next log match after this time, to check whether new sched cycle triggered by runjob from child.

Reply:
self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'}, expect=True) should ensure that the scheduling cycle is finished before returning control back to the test. If that's not happening then we should fix the PTL framework, not this test.

Reply:
There can be a case where on fast machines the sched cycle finishes in the same second as when you set 'now'. So, try adding a sleep(1) after self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'}, expect=True) and before setting 'now'. That should ensure that log_match() only reads the next sched cycle.

Reply:
Yes then that should be useful. I will check for microsecond logging.

Reply:
ah, ok, my bad, i thought it was there. @SudeshnaMoulik i guess you should go the route of adding a sleep(1) before setting 'now', sorry for making you go through a fruitless path.